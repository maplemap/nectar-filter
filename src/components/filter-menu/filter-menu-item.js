import React from 'react';
import {Button} from 'antd';
import {FilterPopup} from 'components/filter-popup';

export const FilterMenuItem = ({
    filterTypes,
    onPopupVisibleChange,
    buttonItemOnClick,
    caption,
    isActive
}) => (
    <div className="filter-menu__item">
        <FilterPopup filterTypes={filterTypes} onVisibleChange={onPopupVisibleChange}>
            <Button {...{
                name: Array.isArray(filterTypes) ? filterTypes.join() : filterTypes,
                onClick: buttonItemOnClick,
                ...(isActive ? {type: 'primary'} : null)
            }}
            >
                {caption}
            </Button>
        </FilterPopup>
    </div>
);
