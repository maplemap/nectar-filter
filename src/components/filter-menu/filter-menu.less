.filter-menu {
  display: flex;

  &__item {
    margin-right: 1rem;
  }
}
