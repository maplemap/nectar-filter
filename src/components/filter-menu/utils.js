import {getCapitalized} from 'utils';

export const getCountCaption = ({name, count}) => {
    const appliedCount = count > 0 ? ` (${count})` : '';
    return `${getCapitalized(name)}${appliedCount}`
};

export const getCountFiltersByType = ({appliedFilters, filterType}) => (
    appliedFilters[filterType] && appliedFilters[filterType].length > 0 ?
        appliedFilters[filterType].length : 0
);
