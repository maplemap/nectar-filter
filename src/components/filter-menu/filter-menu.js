import React, {useState, useContext, useCallback} from 'react';
import {ContextApp} from 'reducer';
import {FilterMenuItem} from './filter-menu-item';
import {getCountCaption, getCountFiltersByType} from './utils'

import './filter-menu.less';

export const FilterMenu = () => {
    const [activeItem, setActiveItem] = useState('');
    const {state: {filtersData, appliedFilters, isMobile}} = useContext(ContextApp);
    const filtersDataKeys = Object.keys(filtersData);

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const buttonItemOnClick = useCallback(e => {
        setActiveItem(e.target.name);
    });

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const onPopupVisibleChange = useCallback(shown => {
        if (!shown) {
            setActiveItem('');
        }
    });

    const getSingleFilterItem = filterType => {
        const count = getCountFiltersByType({appliedFilters, filterType});
        return (
            <FilterMenuItem key={filterType} {...{
                onPopupVisibleChange,
                buttonItemOnClick,
                filterTypes: filterType,
                caption: getCountCaption({count, name: filterType}),
                isActive: activeItem === filterType || count > 0
            }}
            />
        )
    };

    const getMobileMenu = () => {
        const buttonsKeys = filtersDataKeys.slice(0, 2);
        const moreFilterKeys = filtersDataKeys.slice(2);
        const count = (
            moreFilterKeys.reduce((acc, filterType) => (
                acc + getCountFiltersByType({appliedFilters, filterType})
            ), 0)
        );
        return (
            <>
                {buttonsKeys.map(getSingleFilterItem)}
                <FilterMenuItem{...{
                    onPopupVisibleChange,
                    buttonItemOnClick,
                    filterTypes: moreFilterKeys,
                    caption: getCountCaption({count, name: 'More Filters'}),
                    isActive: activeItem === moreFilterKeys.join() || count > 0
                }}
                />
            </>
        )
    };

    return (
        <div className="filter-menu">
            {isMobile && filtersDataKeys.length > 3 ?
                getMobileMenu() :
                Object.keys(filtersData).map(getSingleFilterItem)}
        </div>
    )
};
