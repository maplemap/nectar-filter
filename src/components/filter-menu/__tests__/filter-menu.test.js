import React from 'react';
import {shallow} from 'enzyme';
import {FilterMenu} from '../filter-menu';
import {FilterMenuItem} from '../filter-menu-item';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(),
    useState: jest.fn()
}));

const getComponent = props => (
    <FilterMenu {...{
        ...props
    }}
    />
);
const getShallow = props => shallow(getComponent(props));

describe('FilterMenu', () => {
    const getContextData = ({filtersData = {}, appliedFilters = {}, isMobile = false}) => ({
        state: {filtersData, appliedFilters, isMobile}
    });

    describe('when device is not mobile', () => {
        describe('when `filtersData` has keys less then `3`', () => {
            const size = 'size';
            const filtersData = {[size]: {}};
            let filterMenuItems;
            let setActiveItem;

            beforeEach(() => {
                setActiveItem = jest.fn();
                React.useContext.mockReturnValue(getContextData({filtersData}));
                React.useState.mockImplementation(activeItem => [activeItem, setActiveItem]);
                filterMenuItems = getShallow().find(FilterMenuItem);
            });

            it('should render count of `FilterMenuItem` components according to the count of `filtersData` keys',
                () => {
                    expect(filterMenuItems.length).toBe(Object.keys(filtersData).length);
                });

            it('should render `FilterMenuItem` components according to the `filtersData` keys', () => {
                expect(filterMenuItems.at(0).props().filterTypes).toBe(size);
            });

            describe('when `onPopupVisibleChange` method was invoked', () => {
                describe('when `shown` parameter is equal to `false`', () => {
                    it('should call `setActiveItem` with empty string', () => {
                        filterMenuItems.at(0).props().onPopupVisibleChange(false);

                        expect(setActiveItem).toHaveBeenCalledWith('');
                    });
                })
            });

            describe('when `buttonItemOnClick` method was invoked', () => {
                it('should call `setActiveItem` with current `name` of item', () => {
                    const event = {target: {name: 'name'}};
                    filterMenuItems.at(0).props().buttonItemOnClick(event);

                    expect(setActiveItem).toHaveBeenCalledWith(event.target.name);
                });
            });
        });
    });

    describe('when device is mobile', () => {
        describe('when `filtersData` has keys more then `3`', () => {
            const filtersData = {size: {}, color: {}, room: {}, price: {}};
            let wrapper;
            let filterMenuItems;

            beforeEach(() => {
                React.useContext.mockReturnValue(getContextData({filtersData, isMobile: true}));
                wrapper = getShallow();
                filterMenuItems = wrapper.find(FilterMenuItem);
            });

            it('should render two `FilterMenuItem` components with single value in `filterTypes` property', () => {
                expect(wrapper.find(FilterMenuItem).slice(0, 2)).toMatchSnapshot();
            });

            it('should render one `FilterMenuItem` component with multiple value in `filterTypes` property', () => {
                expect(filterMenuItems.at(2).props().filterTypes).toEqual(Object.keys(filtersData).slice(2));
            });
        });
    })
});
