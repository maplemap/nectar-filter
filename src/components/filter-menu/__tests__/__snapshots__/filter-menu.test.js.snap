// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterMenu when device is mobile when \`filtersData\` has keys more then \`3\` should render two \`FilterMenuItem\` components with single value in \`filterTypes\` property 1`] = `
Array [
  <FilterMenuItem
    buttonItemOnClick={[Function]}
    caption="Size"
    filterTypes="size"
    isActive={false}
    key="size"
    onPopupVisibleChange={[Function]}
  />,
  <FilterMenuItem
    buttonItemOnClick={[Function]}
    caption="Color"
    filterTypes="color"
    isActive={false}
    key="color"
    onPopupVisibleChange={[Function]}
  />,
]
`;
